### DISCUSSION TITLE
Need help with minimum line constraint function in adjoint optimization

### DISCUSSION THREAD

**Discussion by satadrudas** [NanoComp/meep#2698] [Q&A]
Hi,

I am working on a inverse design problem using adjoint optimization. The code structure is similar to that in the [MEEP adjoint solver tutorial](https://meep.readthedocs.io/en/latest/Python_Tutorials/Adjoint_Solver/). I have changed the hyperparameters of the linewidth and linespacing constraint function which activates in the last epoch as given below.

Some relevant parameters:

1. Simulation resolution = 20
2. Design region simulation = 100
3. filter radius = 0.1

```

def glc(result: np.ndarray, x: np.ndarray, gradient: np.ndarray,
        beta: float) -> float:
    """Constraint function for the minimum linewidth.

    Args:
      result: the result of the function evaluation modified in place.
      x: 1d array of size 1+Nx*Ny containing epigraph variable (first
         element) and design weights (remaining elements).
      gradient: the Jacobian matrix with dimensions (1+Nx*Ny,
                num. wavelengths) modified in place.
      beta: bias parameter for projection.

    Returns:
      The value of the constraint function (a scalar).
    """
    t = x[0]  # dummy parameter
    v = x[1:]  # design parameters
    a1 = 1e-4  # hyper parameter (primary)
    b1 = 0  # hyper parameter (secondary)
    gradient[:, 0] = -a1

    filter_f = lambda a: mpa.conic_filter(
        a.reshape(Nx, Ny),
        filter_radius,
        design_region_length,
        design_region_width,
        design_region_resolution,
    )
    threshold_f = lambda a: mpa.tanh_projection(a, beta, eta_i)

    # hyper parameter (constant factor and exponent)
    c0 = 8000   #1e7 * (filter_radius * 1 / resolution) ** 4

    M1 = lambda a: mpa.constraint_solid(a, c0, eta_e, filter_f, threshold_f, 1)
    M2 = lambda a: mpa.constraint_void(a, c0, eta_d, filter_f, threshold_f, 1)

    g1 = grad(M1)(v)
    g2 = grad(M2)(v)

    result[0] = M1(v) - a1 * t - b1
    result[1] = M2(v) - a1 * t - b1

    gradient[0, 1:] = g1.flatten()
    gradient[1, 1:] = g2.flatten()

    t1 = (M1(v) - b1) / a1
    t2 = (M2(v) - b1) / a1

    print(f"glc:, {result[0]}, {result[1]}, {t1}, {t2}")

    return max(t1, t2)

```

also 

```
betas = [8, 16, 32, 64, 128, 256, 512]
max_evals = [80, 80, 80, 100, 100, 100, 100]
```

I have run three different optimizations each with different scenarios for subpixel smoothing and the results are as follows:

1. With subpixel smoothing starting from beta=256
![image](https://github.com/NanoComp/meep/assets/38806771/a4e4ee7e-01eb-477a-8a21-e51cfcfcddcc)

2.With subpixel smoothing starting from beta=64
![image](https://github.com/NanoComp/meep/assets/38806771/f9fb2890-0ce2-41cb-857b-0407ad762061)

3. With no subpixel smoothing
![image](https://github.com/NanoComp/meep/assets/38806771/709385a4-9b8d-4f47-9b37-e9309e094f8d)

Basically, I am facing design degradation in the presence of line constraints (which is expected but I wasn't expecting it to be so heavily degraded) and subpixel smoothing. I have tried several valued for the `c0` hyperparameter but end up with similar results.

Is there any suggestion on how to circumvent this issue?
